File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/clientserver/DeepExample.cs
using System;
using System.IO;
using com.db4o;

namespace com.db4odoc.f1.clientserver
{
	public class DeepExample
    {
		public readonly static string YapFileName = "formula1.yap";

        public static void Main(string[] args)
        {
            File.Delete(YapFileName);
            ObjectContainer db = Db4o.OpenFile(YapFileName);
            try
            {
                StoreCar(db);
                db.Close();
                SetCascadeOnUpdate();
                db = Db4o.OpenFile(YapFileName);
                TakeManySnapshots(db);
                db.Close();
                db = Db4o.OpenFile(YapFileName);
                RetrieveAllSnapshots(db);
                db.Close();
                db = Db4o.OpenFile(YapFileName);
                RetrieveSnapshotsSequentially(db);
                RetrieveSnapshotsSequentiallyImproved(db);
                db.Close();
                SetActivationDepth();
                db = Db4o.OpenFile(YapFileName);
                RetrieveSnapshotsSequentially(db);
            }
            finally
            {
                db.Close();
            }
        }
		// end Main
        
        public static void StoreCar(ObjectContainer db)
        {
            Pilot pilot = new Pilot("Rubens Barrichello", 99);
            Car car = new Car("BMW");
            car.Pilot = pilot;
            db.Set(car);
        }
		// end StoreCar
        
        public static void SetCascadeOnUpdate()
        {
            Db4o.Configure().ObjectClass(typeof(Car)).CascadeOnUpdate(true);
        }
		// end SetCascadeOnUpdate
        
        public static void TakeManySnapshots(ObjectContainer db)
        {
            ObjectSet result = db.Get(typeof(Car));
            Car car = (Car)result.Next();
            for (int i=0; i<5; i++)
            {
                car.Snapshot();
            }
            db.Set(car);
        }
		// end TakeManySnapshots
        
        public static void RetrieveAllSnapshots(ObjectContainer db)
        {
            ObjectSet result = db.Get(typeof(SensorReadout));
            while (result.HasNext())
            {
                Console.WriteLine(result.Next());
            }
        }
		// end RetrieveAllSnapshots
        
        public static void RetrieveSnapshotsSequentially(ObjectContainer db)
        {
            ObjectSet result = db.Get(typeof(Car));
            Car car = (Car)result.Next();
            SensorReadout readout = car.GetHistory();
            while (readout != null)
            {
                Console.WriteLine(readout);
                readout = readout.Next;
            }
        }
		// end RetrieveSnapshotsSequentially
        
        public static void RetrieveSnapshotsSequentiallyImproved(ObjectContainer db)
        {
            ObjectSet result = db.Get(typeof(Car));
            Car car = (Car)result.Next();
            SensorReadout readout = car.GetHistory();
            while (readout != null)
            {
                db.Activate(readout, 1);
                Console.WriteLine(readout);
                readout = readout.Next;
            }
        }
		// end RetrieveSnapshotsSequentiallyImproved
        
        public static void SetActivationDepth()
        {
            Db4o.Configure().ObjectClass(typeof(TemperatureSensorReadout))
                .CascadeOnActivate(true);
        }
		// end SetActivationDepth        
    }
}

File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/clientserver/DeepExample.vb
Imports System
Imports System.IO
Imports com.db4o
Namespace com.db4odoc.f1.clientserver
    Public Class DeepExample
        Public Shared ReadOnly YapFileName As String = "formula1.yap"
        Public Shared Sub Main(ByVal args As String())
            File.Delete(YapFileName)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                StoreCar(db)
                db.Close()
                SetCascadeOnUpdate()
                db = Db4oFactory.OpenFile(YapFileName)
                TakeManySnapshots(db)
                db.Close()
                db = Db4oFactory.OpenFile(YapFileName)
                RetrieveAllSnapshots(db)
                db.Close()
                db = Db4oFactory.OpenFile(YapFileName)
                RetrieveSnapshotsSequentially(db)
                RetrieveSnapshotsSequentiallyImproved(db)
                db.Close()
                SetActivationDepth()
                db = Db4oFactory.OpenFile(YapFileName)
                RetrieveSnapshotsSequentially(db)
            Finally
                db.Close()
            End Try
        End Sub
        ' end Main

        Public Shared Sub StoreCar(ByVal db As ObjectContainer)
            Dim pilot As Pilot = New Pilot("Rubens Barrichello", 99)
            Dim car As Car = New Car("BMW")
            car.Pilot = pilot
            db.[Set](car)
        End Sub
        ' end StoreCar

        Public Shared Sub SetCascadeOnUpdate()
            Db4oFactory.Configure().ObjectClass(GetType(Car)).CascadeOnUpdate(True)
        End Sub
        ' end SetCascadeOnUpdate

        Public Shared Sub TakeManySnapshots(ByVal db As ObjectContainer)
            Dim result As ObjectSet = db.[Get](GetType(Car))
            Dim car As Car = DirectCast(result.[Next](), Car)
            Dim i As Integer = 0
            While i < 5
                car.Snapshot()
                System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
            End While
            db.[Set](car)
        End Sub
        ' end TakeManySnapshots

        Public Shared Sub RetrieveAllSnapshots(ByVal db As ObjectContainer)
            Dim result As ObjectSet = db.[Get](GetType(SensorReadout))
            While result.HasNext()
                Console.WriteLine(result.[Next]())
            End While
        End Sub
        ' end RetrieveAllSnapshots

        Public Shared Sub RetrieveSnapshotsSequentially(ByVal db As ObjectContainer)
            Dim result As ObjectSet = db.[Get](GetType(Car))
            Dim car As Car = DirectCast(result.[Next](), Car)
            Dim readout As SensorReadout = car.GetHistory()
            While Not readout Is Nothing
                Console.WriteLine(readout)
                readout = readout.[Next]
            End While
        End Sub
        ' end RetrieveSnapshotsSequentially

        Public Shared Sub RetrieveSnapshotsSequentiallyImproved(ByVal db As ObjectContainer)
            Dim result As ObjectSet = db.[Get](GetType(Car))
            Dim car As Car = DirectCast(result.[Next](), Car)
            Dim readout As SensorReadout = car.GetHistory()
            While Not readout Is Nothing
                db.Activate(readout, 1)
                Console.WriteLine(readout)
                readout = readout.[Next]
            End While
        End Sub
        ' end RetrieveSnapshotsSequentiallyImproved

        Public Shared Sub SetActivationDepth()
            Db4oFactory.Configure().ObjectClass(GetType(TemperatureSensorReadout)).CascadeOnActivate(True)
        End Sub
        ' end SetActivationDepth
    End Class
End Namespace
